### PR TITLE
Abins: JSON data import

### PR DESCRIPTION
**Pull request by ajjackson**
#### Summary of work
Created a new data import format for Abins: "JSON".

This single JSON Loader can actually identify and import three different relevant JSON formats, and others might be identified in future.

- **AbinsData**: a dump from Abins internal format for phonon/vibration data
- **euphonic.QpointPhononModes**:  an equivalent set of data dummped from the Euphonic library. This would be the recommended tool/format for people who want to modify/customize/generate phonon mode data in Python
- **euphonic.ForceConstants**: A force-constants model from which phonon modes are computed. This is similar to the existing "FORCECONSTANTS" loader which can read *.castep_bin* and *phonopy.yaml* files; the Euphonic library is used to evaluate on a mesh of q-points determined by values in *abins.parameters*

#### Purpose of work
Allow phonon data that has been created or manipulated outside of a supported program to be read in to AbINS.

Currently if a computationally-inclined user has generated or manipulated a set of phonon modes, their best bet is to imitate the CASTEP .phonon format. For experiments with force constants, they would need to export via Phonopy.

We have some other .json formats that are closely related to our projects/tools: Abins internal AbinsData .json dump, and Euphonic QpointPhononModes and ForceConstants .json dumps. Add support for these.

#### Further detail of work
As JSON is a widely-used format, it is quite likely that a user will point this at the wrong file. It is also quite inconvenient to select from a long list of supported formats that have the same extension. So it is quite helpful to detect the file type automatically. Both Abins and Euphonic provide "magic" keys that make their JSON dumps easy to identify, so we can look for these when validating the Algorithm input.

However, JSON files can be huge so decoding the whole thing into memory to do this is not a great idea. The *json-stream* package provides a much more efficient lazy search through JSON files that should avoid laggy performance if the user has a huge file (especially with a lot of top-level entries.) I think this would be a new dependency (already available on conda-forge) and might help performance elsewhere.

So far I set up this dependency optionally but I think we should proabably go ahead and include it in conda deps. How well is that likely to go down?

### To test:

Unit tests have been updates

### Release notes
Are required, but I don't think there is anywhere to put them yet?

---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

### CODE REVIEW

**Comment by ajjackson** [comment]
That's quite a nasty test failure: when importing force constants, the eigenvectors differ - but only on Windows!

Whereas most Abins formats already contain displacement values (and we expect exact agreement), here they are computed using Euphonic from the force constants. There are multiple solutions to the set of eigenvectors; looking at the numbers here they are showing up as differences in sign (i.e. direction; phase, really). Comparing eigenvectors is known to be hard. But it's quite annoying that it manifests as a platform dependence... At heart it could be a sensitivity to the compiler or numerical libraries.

I assume this is why the AbinsLoadPhonopyTest does not compare with reference data. But we should be able to do slightly better; the frequencies agree, after all. I will introduce logic to skip the displacements check for this test.

**Review by SilkeSchomann — DISMISSED** [review]
The new code for JSON data looks good and is covered by unit tests.

**Comment by SilkeSchomann** [comment]
I am aware that there are conflicts to be resolved, but the changes are still ready to be approved.

**Comment by ajjackson** [comment]
@SilkeSchomann I'm not sure what the failure is about, may to do with permissions to remove the merge conflict label?

**Review by SilkeSchomann — APPROVED** [review]
Happy to approve again after the merge conflict got fixed.

File: scripts/abins/input/jsonloader.py
Comment (robertapplin) [inline scripts/abins/input/jsonloader.py:63]: Interesting, never seen this before! Was it introduced in a recent python version?

**Pull request merged**
